### PR TITLE
Completed several features for Weather App

### PR DESCRIPTION
**Pull request by SpartaSixZero**
# Weather API Technical Assessment

**Note: Please ensure that location services are enabled for your browser of choice.**

## Features completed:

 - As a user, I would like to search for the current weather in my area
 - As a user, I would like to see the 5-day forecast
 - As a user, I would like to toggle between celsius and fahrenheit

## Features not completed:
 - As a user, I would like to share the URL of the weather in my area and see the results

Note: For this feature above, I would feel it would be best to discuss with the originator of the requirements to get a better understanding of the goal before I can come up with a solution that makes sense.

Also, with more time, I can focus on adding some unit tests.

## Screenshots

Screenshot of Current Weather:
image

Screenshot of 5 day forecast:
image

## Summary
1. Using browser to retrieve your current location in order to get latitude and longitude to get weather data.
1. Added a Menu bar which allows the user to toggle between Current Weather module and 5 day forecast module
1. Current Weather module displays current temperature while 5 day forecast display daily high and daily low
1. Added Radio checkbox to allow for easily toggling between Fahrenheit and Celsius.
1. Leveraged React Hooks so we only have React function components.
1. Leveraged Axios to make API calls
1. Leveraged Semantic UI for some basic styling

### CODE REVIEW

File: src/components/app.jsx [hunk 1]
```diff
@@ -1,9 +1,74 @@
...
+  const [lat, setLat] = useState(null);
+  const [lon, setLon] = useState(null);
+  const [err, setErr] = useState(null);
+  const [unit, setUnit] = useState("imperial");
+
+  const OPEN_WEATHER_APP_ID = "62536724de969bf01c1e02a249c958f9";
+
+  useEffect(() => {
+    // Make a one time call to our browser to get current position
+    window.navigator.geolocation.getCurrentPosition(
```
Comment (jlaurelli) [inline src/components/app.jsx]: What made you go with this solution as opposed to something like a search bar?

File: src/components/app.jsx [hunk 2]
```diff
@@ -1,9 +1,74 @@
...
+
+    if (!lon || !lat) {
+      return (
+        <div class="ui active dimmer">
+          <div class="ui text loader">Please wait...</div>
+        </div>
+      );
+    }
+
+    if (currentView === "currentWeather") {
```
Comment (jlaurelli) [inline src/components/app.jsx:21]: Since changing options on the `Nav` will update the `currentView` value, each of these components will be mounted/unmounted when switched, causing a new api call.

What are the benefits of this? What problems do you foresee with this approach?

**Review by jlaurelli — COMMENTED** [review]
Hi Mark! I'm James, one of the engineers at Contently. I left a few comments about your decisions. If you have time to respond to them, that'd be great!

**Comment by jlaurelli** [comment]
If you were write tests for your code here, how would you go about doing that? Feel free to discuss what libraries (e.g. `jest`, `enzyme`, etc.) you like or strategies (`unit`, `integration`, etc.) you prefer.

File: src/components/app.jsx (hunk 1 above)
Comment (SpartaSixZero) [reply src/components/app.jsx]: Hi James,

When I first saw this requirement:
 _As a user, I would like to search for the current weather in my area_

I immediately remembered that just recently, I've learned and used the Geolocation API in another app. I thought that it would definitely save users the trouble of having to type their current city into a search bar and save some time. However, the drawback with this approach is that currently, this app is only limited to your current location. I think a good future enhancement would be to add a search bar so that users can see current weather and 5 day forecast of other places.

File: src/components/app.jsx (hunk 2 above)
Comment (SpartaSixZero) [reply src/components/app.jsx:21]: I can see the benefit of keeping data in CurrentWeather component refreshed. For example, if a user was looking at this CurrentWeather view at 8am and then switched to 5 Day view and went off to do something else for a while, if that user then switches back to CurrentWeather let's say at 8:30am or 9am, then the current weather temperature could have changed. I can't say if that is a common use case.

A possible problem could be that since 5 day forecast, since it only shows daily Highs and Lows, should not change given the same days, it's not that efficient to make a new network request every time we mount the component. A possible workaround is to add some sort of caching or lazy loading of the 5 day forecast data. Here, I'm also assuming location never changes.

**Comment by SpartaSixZero** [comment]
As for tests, I would lean on unit tests using `jest` and `enzyme`.

For me, I feel most natural starting with unit testing the smallest components first. Starting with Nav component first.
## Nav
- I would create a test to check that Nav component renders correctly
  - A Menu component is rendered. That Menu component has two Menu.Item for children
  - One Menu.Item displays "Current Weather"
  - The second Menu.Item displays "5 Day"
- I would create another test to ensure that the click handler is working
  - I would use Enzyme simulate() to simulate a click event on both Menu.Item components and verify that the onClickHandler is getting invoked with the correct parameters.
  - I would also test that onViewChange was invoked if we are on one component and the Nav option for the other component was clicked.

## UnitCheckbox
- Similar to Nav, I would create a test(s) to ensure that the component is rendered correctly
- I would also create another test to ensure that onChange is invoked if one radio input is selected and the other input is clicked.

## WeatherCard
- Since this component can differ depending what props are passed in, I can create different props to easily switch between different states. This can help me test different variations.

## CurrentWeather
- Here, since we're making a network call with Axios, I would pass in some fake parameters as props and instead of making an actual API call, I'll mock it and return a fake response data. The reason is that we want our unit tests to be fast and network calls can be slow. Once our canned response comes back, we can test that the correct functions for updating our State gets called.

## FiveDayForecast
- Similar to CurrentWeather in terms of faking the API call.

## app
- Here, I want to mock out `window.navigator.geolocation.getCurrentPosition` and have different tests for different scenarios. For example, I can have 1 test for a success scenario and another test where it failed in order to test the error callback. 
- I can also have a test to set lat and lon to null to test the loading indicator is rendering properly.
- For rendering
  - I want a test that Nav element is present along with the UnitCheckbox component.
  - I can set different state values for currentView to ensure that either CurrentWeather or FiveDayForecast is also rendered.

**Comment by jlaurelli** [comment]
Awesome! Thanks for the clarification!

**Comment by SpartaSixZero** [comment]
Np! Glad I can help!
